hazelcast/src/main/java/com/hazelcast/spi/impl/executionservice/impl/ExecutionServiceImpl.java

```diff
@@ -316,34 +318,45 @@ public TaskScheduler getTaskScheduler(String name) {
     public void shutdown() {
         logger.finest("Stopping executors...");
         scheduledExecutorService.notifyShutdownInitiated();
-        for (ExecutorService executorService : executors.values()) {
-            executorService.shutdown();
-        }
-        for (ExecutorService executorService : durableExecutors.values()) {
-            executorService.shutdown();
-        }
-        for (ExecutorService executorService : scheduleDurableExecutors.values()) {
-            executorService.shutdown();
-        }
+
+        shutdown(executors);
+        shutdown(durableExecutors);
+        shutdown(scheduleDurableExecutors);
         scheduledExecutorService.shutdownNow();
         cachedExecutorService.shutdown();
-        try {
-            scheduledExecutorService.awaitTermination(AWAIT_TIME, TimeUnit.SECONDS);
-        } catch (InterruptedException e) {
-            currentThread().interrupt();
-            logger.finest(e);
+
+        awaitAndForceShutdown(executors);
+        awaitAndForceShutdown(scheduledExecutorService);
+        awaitAndForceShutdown(cachedExecutorService);
+
+        executors.clear();
+        durableExecutors.clear();
+        scheduleDurableExecutors.clear();
+    }
+
+    private void shutdown(Map<?, ? extends ExecutorService> executorServiceMap) {
+        for (ExecutorService executorService : executorServiceMap.values()) {
+            executorService.shutdown();
         }
+    }
+
+    private void awaitAndForceShutdown(Map<?, ? extends ExecutorService> executorServiceMap) {
+        for (ExecutorService executorService : executorServiceMap.values()) {
+            awaitAndForceShutdown(executorService);
+        }
+    }
+
+    private void awaitAndForceShutdown(ExecutorService executorService) {
         try {
-            if (!cachedExecutorService.awaitTermination(AWAIT_TIME, TimeUnit.SECONDS)) {
-                cachedExecutorService.shutdownNow();
+            if (!executorService.awaitTermination(AWAIT_TIME, TimeUnit.MILLISECONDS)) {
+                executorService.shutdownNow();
             }
         } catch (InterruptedException e) {
             currentThread().interrupt();
             logger.finest(e);
+        } catch (UnsupportedOperationException e) {
```

Comment:
i think some executors are not supporting it, if it is the case, maybe you can add a small line comment for it.